resolver/ipinfo.go

```diff
@@ -150,27 +152,38 @@ func GetIPInfo(ip string) (*IPInfo, error) {
 }
 
 // Save saves the IPInfo record to the database.
-func (ipi *IPInfo) Save() error {
-	ipi.Lock()
-	if !ipi.KeyIsSet() {
-		ipi.SetKey(makeIPInfoKey(ipi.IP))
+func (info *IPInfo) Save() error {
+	info.Lock()
+
+	// Set database key if not yet set already.
+	if !info.KeyIsSet() {
+		// Default to global scope if scope is unset.
+		if info.Scope == "" {
+			info.Scope = IPInfoScopeGlobal
+		}
+		info.SetKey(makeIPInfoKey(info.Scope, info.IP))
 	}
-	ipi.Unlock()
-
-	// Legacy support
-	// Ensure we don't write new Domain fields into the
-	// database.
-	// DEPRECATED: remove with alpha
-	if len(ipi.Domains) > 0 {
-		ipi.Domains = nil
+
+	// Calculate and set cache expiry.
+	var expires int64 = 86400 // Minimum TTL of one day.
+	for _, rd := range info.ResolvedDomains {
+		if rd.Expires > expires {
+			expires = rd.Expires
+		}
 	}
+	info.UpdateMeta()
+	expires += 3600 // Add one hour to expiry as a buffer.
+	info.Meta().SetAbsoluteExpiry(expires)
+
+	info.Unlock()
 
-	return ipInfoDatabase.Put(ipi)
+	return ipInfoDatabase.Put(info)
 }
```

Comment:
`info` should be locked when calling `Put` so defer the unlock at the beginning

Reply:
Ah, no. `Put` actually does lock the record.